soccer/planning/SettlePathPlanner.cpp

```diff
@@ -0,0 +1,419 @@
+    RJ::Seconds timeIntoPreviousPath;
+
+    // How much of the future we are devoting to the partial path
+    // 0ms unless we have a partial path, then it's partialReplanLeadTime
+    RJ::Seconds partialPathTime = 0ms;
+
+    // How much of the previous path to steal
+    const RJ::Seconds partialReplanLeadTime = RRTPlanner::getPartialReplanLeadTime();
+
+    // Max angle from the ball vector when trying to bounce the ball
+    // on dampen when trying to speed up actions after captures
+    const float maxAngleoffBallForDampen = 45; // deg
+
+    // Search settings when trying to find the correct intersection location
+    // between the fast moving ball and our collecting robot
+    const RJ::Seconds searchStartTime = RJ::Seconds(*_searchStartTime);
+    const RJ::Seconds searchEndTime = RJ::Seconds(*_searchEndTime);
+    const RJ::Seconds searchIncTime = RJ::Seconds(*_searchIncTime);
+
+    const RJ::Seconds bufferInterceptTime = RJ::Seconds(*_interceptBufferTime);
+
+    const float maxAnglePathTargetChange = *_maxAnglePathTargetChange*M_PI/180.0f;
+    const float maxBallAngleChangeForPathReset = *_maxBallAngleForReset*M_PI/180.0f;
+
+    // If the ball changed directions or magnitude really quickly, do a reset of target
+    // But if we are already in another state, it should be very quickly to jump out
+    // and try again
+    if (averageBallVel.angleBetween(ball.vel) > maxBallAngleChangeForPathReset ||
+        (averageBallVel - ball.vel).mag() > *_maxBallVelForPathReset) {
+        firstTargetPointFound = false;
+        firstBallVelFound = false;
+    }
+
+
+    // Smooth out the ball velocity a little bit so we can get a better estimate of intersect points
+    if (firstBallVelFound) {
+        averageBallVel = *_ballVelGain * ball.vel +
+                         (1 - *_ballVelGain) * averageBallVel;
+    } else {
+        averageBallVel = ball.vel;
+        firstBallVelFound = true;
+    }
+
+    // Change start instant to be the partial path end instead of the robot current location
+    // if we actually have already calculated a path the frame before
+    if (prevPath) {
+        timeIntoPreviousPath = curTime - prevPath->startTime();
+
+        // Make sure we still have time in the path to replan and correct
+        // since it's likely that the old path is slightly off
+        //
+        // ---|----------------|-----------------|
+        // TimeNow     EndPartialPrevPath  FinalTargetPoint
+        //                     |-----------------|
+        //          Amount of the path we can change this iteration
+        if (timeIntoPreviousPath < prevPath->getDuration() - partialReplanLeadTime &&
+            timeIntoPreviousPath > 0ms) {
+            partialPath =
+                prevPath->subPath(0ms, timeIntoPreviousPath + partialReplanLeadTime);
+            partialPathTime = partialPath->getDuration() - timeIntoPreviousPath;
+            startInstant = partialPath->end().motion;
+        }
```

Comment:
Is this copy-pasted from the earlier one? If so it might be good to refactor some of this (Also, if so, why is `partialReplanLeadTime` multiplied by 2 in the other one and not this one?)

Reply:
The should both be the same, rip copy paste.